File: UltraMesh 2d animation/Core/Tools/SkewTool.swift
import Foundation
import Metal
import simd

final class SkewTool: Tool {
    let type: ActiveTool = .skew

    private var activeID: UUID?
    private var activeBoneID: UUID?
    private var startSkew = SIMD2<Float>(repeating: 0)
    private var activeAxis: ShearAxis?
    private var startAngle: Float = 0
    private var boneCenter = SIMD2<Float>(repeating: 0)
    private let skewState: SkewGizmoState
    private var mouseDownPosition: SIMD2<Float>?
    private var didDrag = false

    private let dragThreshold: Float = 4

    init(skewState: SkewGizmoState) {
        self.skewState = skewState
    }

    private func ensureDragStarted(currentPosition: SIMD2<Float>) -> Bool {
        if didDrag { return true }
        guard let down = mouseDownPosition else { return false }
        if simd_distance(currentPosition, down) >= dragThreshold {
            didDrag = true
            return true
        }
        return false
    }

    func onMouseDown(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        guard case let .skewEdge(index) = input.activeHandle else { return }
        scene.beginInteraction()
        mouseDownPosition = input.position
        didDrag = false
        if let boneID = scene.selectedBoneID,
           let segment = scene.skeleton.lineSegment(for: boneID) {
            activeBoneID = boneID
            startSkew = scene.skeleton.bones[boneID]?.localTransform.skew ?? .zero
            activeAxis = switch index {
            case 0: .shearX
            case 1: .shearY
            default: .shearZ
            }
            boneCenter = (segment.start + segment.end) * 0.5
            startAngle = atan2(input.position.y - boneCenter.y, input.position.x - boneCenter.x)
            return
        }
        let hit = scene.selectedImageID ?? ToolUtilities.hitTestScreen(
            screenPoint: input.screenPosition,
            viewSize: input.viewSize,
            scene: scene,
            assets: assets,
            camera: input.camera
        )
        guard let id = hit,
              let image = scene.image(for: id),
              let camera = input.camera else { return }
        scene.selectedImageID = id
        activeID = id
        startSkew = image.skew
        activeAxis = switch index {
        case 0: .shearX
        case 1: .shearY
        default: .shearZ
        }
        skewState.shearX = image.skew.x
        skewState.shearY = image.skew.y
        let centerScreen = camera.worldToScreen(image.position, viewSize: input.viewSize)
        startAngle = atan2(input.screenPosition.y - centerScreen.y, input.screenPosition.x - centerScreen.x)
        let hitArc = index + 1
        skewState.mouseDown(screenPt: CGPoint(x: CGFloat(input.screenPosition.x), y: CGFloat(input.screenPosition.y)), hitArc: hitArc)
    }

    func onMouseDrag(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        guard ensureDragStarted(currentPosition: input.position) else { return }
        if let boneID = activeBoneID,
           let axis = activeAxis {
            let currentAngle = atan2(input.position.y - boneCenter.y, input.position.x - boneCenter.x)
            var d = (startAngle - currentAngle) * 180 / .pi
            while d > 180 { d -= 360 }
            while d < -180 { d += 360 }
            if input.isShiftPressed {
                d = d.rounded()
            }
            var next = startSkew
            switch axis {
            case .shearX:
                next.x = (startSkew.x + d).clamped(to: -180...180)
            case .shearY:
                next.y = (startSkew.y + d).clamped(to: -180...180)
            case .shearZ:
                next.x = (startSkew.x + d).clamped(to: -180...180)
                next.y = (startSkew.y - d).clamped(to: -180...180)
            }
            scene.setBoneSkew(id: boneID, skew: next)
            return
        }
        guard let id = activeID,
              let axis = activeAxis,
              let image = scene.image(for: id),
              let camera = input.camera else { return }
        let centerScreen = camera.worldToScreen(image.position, viewSize: input.viewSize)
        let currentAngle = atan2(input.screenPosition.y - centerScreen.y, input.screenPosition.x - centerScreen.x)
        var d = (startAngle - currentAngle) * 180 / .pi
        while d > 180 { d -= 360 }
        while d < -180 { d += 360 }
        if input.isShiftPressed {
            d = d.rounded()
        }
        var next = startSkew
        switch axis {
        case .shearX:
            next.x = (startSkew.x + d).clamped(to: -180...180)
        case .shearY:
            next.y = (startSkew.y + d).clamped(to: -180...180)
        case .shearZ:
            next.x = (startSkew.x + d).clamped(to: -180...180)
            next.y = (startSkew.y - d).clamped(to: -180...180)
        }

        scene.setImageSkew(id: id, skew: next)

        skewState.applyDrag(
            deltaDegrees: d,
            axis: axis,
            isSnap: input.isShiftPressed,
            cursor: CGPoint(x: CGFloat(input.screenPosition.x), y: CGFloat(input.screenPosition.y))
        )
    }

    func onMouseUp(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        scene.endInteraction()
        defer {
            activeBoneID = nil
            activeID = nil
            activeAxis = nil
            mouseDownPosition = nil
            didDrag = false
            skewState.mouseUp()
        }
        guard didDrag else { return }
        if let boneID = activeBoneID,
           let bone = scene.skeleton.bones[boneID] {
            scene.commitKeyframe(for: boneID, property: .shear, value: .shear(bone.localTransform.skew))
            return
        }
        if let id = activeID, let image = scene.image(for: id) {
            scene.commitKeyframe(for: id, property: .shear, value: .shear(image.skew))
        }
    }

    func update(scene: SceneManager, assets: AssetManager) {
        guard activeID == nil, activeBoneID == nil else { return }
        guard let selectedID = scene.selectedImageID, let image = scene.image(for: selectedID) else { return }
        if skewState.shearX != image.skew.x || skewState.shearY != image.skew.y {
            skewState.shearX = image.skew.x
            skewState.shearY = image.skew.y
        }
    }
}

File: UMeshCore/src/Editor/Tools/SkewTool.cpp
#include "umeshcore/Editor/Tools/SkewTool.h"

#include <algorithm>
#include <cmath>
#include <variant>

#include "umeshcore/Math/MatrixUtilities.h"

namespace umeshcore {

bool SkewTool::ensureDragStarted(const Vec2& currentPosition) {
    if (didDrag_) return true;
    if (!mouseDownPosition_.has_value()) return false;
    if (length(currentPosition - *mouseDownPosition_) >= kDragThreshold) {
        didDrag_ = true;
        return true;
    }
    return false;
}

std::optional<ShearAxis> SkewTool::axisFor(const GizmoHandle& handle) {
    const auto* edge = std::get_if<SkewEdgeHandle>(&handle);
    if (edge == nullptr) return std::nullopt;
    if (edge->index == 0) return ShearAxis::ShearX;
    if (edge->index == 1) return ShearAxis::ShearY;
    return ShearAxis::ShearZ;
}

namespace {
float normalizedDegrees(float d) {
    while (d > 180.0f) d -= 360.0f;
    while (d < -180.0f) d += 360.0f;
    return d;
}

Vec2 appliedSkew(ShearAxis axis, const Vec2& startSkew, float d) {
    Vec2 next = startSkew;
    switch (axis) {
        case ShearAxis::ShearX:
            next.x = std::clamp(startSkew.x + d, -180.0f, 180.0f);
            break;
        case ShearAxis::ShearY:
            next.y = std::clamp(startSkew.y + d, -180.0f, 180.0f);
            break;
        case ShearAxis::ShearZ:
            next.x = std::clamp(startSkew.x + d, -180.0f, 180.0f);
            next.y = std::clamp(startSkew.y - d, -180.0f, 180.0f);
            break;
    }
    return next;
}
} // namespace

void SkewTool::onMouseDown(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& imageHitTest, float /*hitScale*/,
    bool /*touchOptimized*/) {
    if (!input.activeHandle.has_value()) return;
    const auto axis = axisFor(*input.activeHandle);
    if (!axis.has_value()) return;

    scene.beginInteraction();
    mouseDownPosition_ = input.position;
    didDrag_ = false;

    if (scene.selectedBoneID.has_value()) {
        const auto segment = scene.skeleton.lineSegment(*scene.selectedBoneID);
        if (segment.has_value()) {
            activeBoneID_ = scene.selectedBoneID;
            const Bone* bone = scene.skeleton.bone(*scene.selectedBoneID);
            startSkew_ = bone != nullptr ? bone->localTransform.skew : Vec2::zero();
            activeAxis_ = axis;
            boneCenter_ = (segment->start + segment->end) * 0.5f;
            startAngle_ = std::atan2(input.position.y - boneCenter_.y, input.position.x - boneCenter_.x);
            return;
        }
    }

    const std::optional<ImageHit> imageHit =
        imageHitTest ? imageHitTest(input.screenPosition, input.viewSize, input.camera) : std::nullopt;
    const std::optional<Uuid> hit =
        scene.selectedImageID.has_value() ? scene.selectedImageID
        : (imageHit.has_value() ? std::optional<Uuid>(imageHit->id) : std::nullopt);
    if (!hit.has_value()) return;
    SceneImage* image = scene.image(*hit);
    if (image == nullptr || input.camera == nullptr) return;

    scene.selectedImageID = *hit;
    activeID_ = *hit;
    startSkew_ = image->skew;
    activeAxis_ = axis;
    const Vec2 centerScreen = input.camera->worldToScreen(image->position, input.viewSize);
    startAngle_ =
        std::atan2(input.screenPosition.y - centerScreen.y, input.screenPosition.x - centerScreen.x);
}

void SkewTool::onMouseDrag(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool /*touchOptimized*/) {
    if (!ensureDragStarted(input.position)) return;

    if (activeBoneID_.has_value() && activeAxis_.has_value()) {
        const float currentAngle =
            std::atan2(input.position.y - boneCenter_.y, input.position.x - boneCenter_.x);
        float d = normalizedDegrees((startAngle_ - currentAngle) * 180.0f / kPi);
        if (input.isShiftPressed) d = std::round(d);
        scene.setBoneSkew(*activeBoneID_, appliedSkew(*activeAxis_, startSkew_, d));
        return;
    }

    if (!activeID_.has_value() || !activeAxis_.has_value() || input.camera == nullptr) return;
    SceneImage* image = scene.image(*activeID_);
    if (image == nullptr) return;

    const Vec2 centerScreen = input.camera->worldToScreen(image->position, input.viewSize);
    const float currentAngle =
        std::atan2(input.screenPosition.y - centerScreen.y, input.screenPosition.x - centerScreen.x);
    float d = normalizedDegrees((startAngle_ - currentAngle) * 180.0f / kPi);
    if (input.isShiftPressed) d = std::round(d);
    scene.setImageSkew(*activeID_, appliedSkew(*activeAxis_, startSkew_, d));
}

void SkewTool::onMouseUp(
    const ToolInput& /*input*/, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool /*touchOptimized*/) {
    scene.endInteraction();

    const std::optional<Uuid> boneID = activeBoneID_;
    const std::optional<Uuid> imageID = activeID_;
    const bool didDrag = didDrag_;

    activeBoneID_ = std::nullopt;
    activeID_ = std::nullopt;
    activeAxis_ = std::nullopt;
    mouseDownPosition_ = std::nullopt;
    didDrag_ = false;

    if (!didDrag) return;

    if (boneID.has_value()) {
        const Bone* bone = scene.skeleton.bone(*boneID);
        if (bone != nullptr) {
            scene.commitKeyframe(*boneID, AnimationTrackProperty::Shear, ShearValue{bone->localTransform.skew});
        }
        return;
    }
    if (imageID.has_value()) {
        const SceneImage* image = scene.image(*imageID);
        if (image != nullptr) {
            scene.commitKeyframe(*imageID, AnimationTrackProperty::Shear, ShearValue{image->skew});
        }
    }
}

} // namespace umeshcore
